src/common/meta/src/wal.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub mod kafka;
+pub mod options_allocator;
+
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+use serde_with::with_prefix;
+
+use crate::error::Result;
+use crate::wal::kafka::KafkaConfig;
+pub use crate::wal::kafka::{KafkaOptions as KafkaWalOptions, Topic as KafkaWalTopic};
+pub use crate::wal::options_allocator::WalOptionsAllocator;
+
+// An encoded wal options will be wrapped into a (WAL_OPTIONS_KEY, encoded wal options) key-value pair
+// and inserted into the options of a `RegionCreateRequest`.
+pub const WAL_OPTIONS_KEY: &str = "wal_options";
+
+/// Wal configurations for bootstrapping meta srv.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, Default)]
+#[serde(tag = "provider")]
+pub enum WalConfig {
+    #[default]
+    #[serde(rename = "raft-engine")]
+    RaftEngine,
+    #[serde(rename = "kafka")]
+    Kafka(KafkaConfig),
+}
+
+/// Wal options allocated to a region.
+// A wal options is encoded by meta srv into a `String` with `serde_json::to_string`.
+// It's then decoded by datanode to a `HashMap<String, String>` with `serde_json::from_str`.
+// Such a encoding/decoding scheme is inspired by the encoding/decoding of `RegionOptions`.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
+#[serde(tag = "wal.provider")]
+pub enum WalOptions {
+    #[default]
+    #[serde(rename = "raft-engine")]
+    RaftEngine,
+    #[serde(rename = "kafka")]
+    #[serde(with = "prefix_wal_kafka")]
+    Kafka(KafkaWalOptions),
+}
```

Comment:
Prefer to don't mix the case styles.
```suggestion
pub enum WalOptions {
    #[default]
    #[serde(rename = "raft_engine")]
    RaftEngine,
    #[serde(rename = "kafka", with = "prefix_wal_kafka")]
    Kafka(KafkaWalOptions),
}
```

Reply:
`prefix_wal_kafka` is a function generated by `with_prefix` macro defined in `serde_with` crate.

Reply:
Changed to snake case.